compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp

```diff
@@ -41,6 +46,315 @@ void mlir::iree_compiler::registerLinalgTransformDialectExtension(
+    // clang-format on
+    rewriter.eraseOp(parallelInsertOp);
+  }
+}
+
+/// Rewrite ExtractSlice ops in `dispatchOp` as Flow::DispatchTensorLoadOps.
+/// Takes a list of all tensor and all tensorDynamicDims operands to the
+/// dispatchOp as well as a BlockAndValueMapping from tensor operands to the
+/// corresponding Flow dispatch tensor bbArgs.
+static void rewriteExtractSlices(PatternRewriter &rewriter,
+                                 Flow::DispatchWorkgroupsOp dispatchOp,
+                                 ValueRange tensorOperands,
+                                 ValueRange tensorDynamicDims,
+                                 BlockAndValueMapping tensorToFlowBvm) {
+  dispatchOp.walk([&](tensor::ExtractSliceOp extractSliceOp) {
+    Value source = extractSliceOp.source();
+    auto it = llvm::find(tensorOperands, source);
+    if (it == tensorOperands.end()) return;
+    int64_t index = std::distance(tensorOperands.begin(), it);
+    Value sourceFlow = tensorToFlowBvm.lookupOrNull(source);
+    if (!sourceFlow) return;
+
+    Location loc = extractSliceOp.getLoc();
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(extractSliceOp);
+    auto dynamicDims =
+        Util::findVariadicDynamicDims(index, tensorOperands, tensorDynamicDims);
+    // clang-format off
+    Value load = rewriter.create<Flow::DispatchTensorLoadOp>(
+        loc,
+        sourceFlow,
+        dynamicDims,
+        extractSliceOp.getMixedOffsets(),
+        extractSliceOp.getMixedSizes(),
+        extractSliceOp.getMixedStrides());
+    // clang-format on
+    rewriter.replaceOp(extractSliceOp, load);
+  });
+}
+
+/// Pattern to rewrite a ForeachThreadOp into a Flow::DispatchWorkGroupsOp.
+/// This rewrite proceeds in a few steps:
+///   - Step 1: Compute the result types and their result dynamic dim
+///   operands.
+///     This first step takes advantage of the ops contained in the
+///     ForeachThreadOp terminator and that are tied to the results.
+///   - Step 2: Get values defined above and separate them between
+///   non-tensors,
+///     tensors and introduce appropriate tensor dims.
+///   - Step 3: Compute workgroupCount.
+///   - Step 4: Create ordered vectors of operands to pass to the builder and
+///     build the dispatchOp.
+///   - Step 5: Fixup dispatchOp bbArgs and terminator.
+///   - Step 6: Move the body of foreachThreadOp to the dispatchOp.
+///   - Step 7: Set up bvm for RAUWIf. In particular, tensor operands become
+///     flow dispatch tensor bbArgs and need to be
+///     flow.dispatch.tensor.load'ed.
+///   - Step 8: Plug dispatch workgroup id and count values into the bvm.
+///   - Step 9. Rewrite tensor::ExtractSlice and ParallelInsert ops to the
+///     relevant Flow DispatchTensorLoad/Store version.
+///   - Step 10: Perform RAUWIf.
+///   - Step 11: Drop the terminator and replace foreachThreadOp.
+// TODO: n-D ForeachThreadOp
+struct ForeachThreadToFlowDispatchWorkgroupsRewriter
```

Comment:
This is to avoid the complexity of setting up another set of files for the purpose of registering a single op in the transform dialect. I can add those as a follow up.